### PR TITLE
hashes:: Rename const_hash functions

### PR DESCRIPTION
**Pull request by tcharding**
There are a number of issues with the two `const_hash` functions in the `sha256` module:

- The two `const_hash` functions in the `sha256` module differ slightly, one finalizes the hash and one is for computing the midstate.
- They are inefficient and provided for usage for const context only.

Fix both issues by renaming the functions as discussed in #3075.

Close: #3075

### CODE REVIEW

File: hashes/src/sha256.rs [hunk 1]
```diff
@@ -336,7 +344,7 @@ impl Midstate {
         w
     }
 
-    const fn const_hash(bytes: &[u8], finalize: bool) -> Self {
+    const fn compute_midstate_unoptimized(bytes: &[u8], finalize: bool) -> Self {
```
Comment (Kixunil) [inline hashes/src/sha256.rs:347]: This is not really a great name because it can also output hash. Maybe `perform_hashing_unoptimized`?

**Review by Kixunil — APPROVED** [review]
ACK e7762e0612ce38cb2f934ccf91822003a71de096

**Comment by apoelstra** [comment]
@Kixunil in your ACK you dropped the leading `e` of the commit ID. I edited your approval message to add it. I believe this was your intention but let me know if I overstepped.

**Review by apoelstra — APPROVED** [review]
ACK e7762e0612ce38cb2f934ccf91822003a71de096 successfully ran local tests

**Pull request merged**